lib/Dialect/ESI/runtime/cpp/include/esi/Accelerator.h

```diff
@@ -23,30 +23,44 @@
 
 #include <cstdint>
 #include <functional>
+#include <map>
 #include <memory>
 #include <string>
+#include <typeinfo>
 
 namespace esi {
-class SysInfo;
 
-/// An ESI accelerator system.
-class Accelerator {
-public:
-  virtual ~Accelerator() = default;
+constexpr uint32_t MagicNumOffset = 16;
+constexpr uint32_t MagicNumberLo = 0xE5100E51;
+constexpr uint32_t MagicNumberHi = 0x207D98E5;
+constexpr uint32_t VersionNumberOffset = MagicNumOffset + 8;
+constexpr uint32_t ExpectedVersionNumber = 0;
 
-  virtual const SysInfo &sysInfo() = 0;
+namespace services {
+/// Parent class of all APIs modeled as 'services'. May or may not map to a
+/// hardware side 'service'.
+class Service {
+public:
+  using Type = const std::type_info &;
+  virtual ~Service() = default;
 };
+} // namespace services
 
-/// Information about the Accelerator system.
-class SysInfo {
+/// An ESI accelerator system.
+class Accelerator {
 public:
-  virtual ~SysInfo() = default;
+  virtual ~Accelerator() = default;
 
-  /// Get the ESI version number to check version compatibility.
-  virtual uint32_t esiVersion() const = 0;
+  template <typename ServiceClass>
+  ServiceClass *getService() {
+    return dynamic_cast<ServiceClass *>(getServiceImpl(typeid(ServiceClass)));
+  }
 
-  /// Return the JSON-formatted system manifest.
-  virtual std::string rawJsonManifest() const = 0;
+protected:
+  using Service = services::Service;
+  virtual Service *createService(Service::Type service) = 0;
```

Comment:
Please use smart pointers here or write a comment to be very explicit about the ownership of the returned pointer (i see a mix of `unique_ptr`s being placed into caches and raw-pointers being allocated and return to the former).

Reply:
This is a protected method because it is not intended to be called directly by clients. It is called by `getServiceImpl`, which immediately wraps it in a unique pointer and adds it to the cache. I structured it that way to avoid subclasses from having to reason about the way that the cache is implemented. I will add some documentation to that effect.